# Session 1
## Experiment 3
### Home work -- Optionally @ Lab

### Data Source
In this experiment, we will use famous Iris data set.
https://archive.ics.uci.edu/ml/datasets/iris

### Objective
To use the kNN classifier for multi-class classification.

#### Dataset Information:

This is perhaps the best known database to be found in the pattern recognition literature. Fisher's paper is a classic in the field and is referenced frequently to this day. 

The data set contains 3 classes of 50 instances each, where each class refers to a type of iris plant. One class is linearly separable from the other 2; the latter are NOT linearly separable from each other. 


### Data Attributes

  1. sepal length in cm 
  2. sepal width in cm 
  3. petal length in cm 
  4. petal width in cm 
  5. class: 
     -- Iris Setosa  
     -- Iris Versicolour 
     -- Iris Virginica

### Predicted attribute
class of iris plant. 

In [1]:
import pandas as pd
pd.read_csv("../../Datasets/bezdekIris.data", header=None)

0    1    2    3               4
0    5.1  3.5  1.4  0.2     Iris-setosa
1    4.9  3.0  1.4  0.2     Iris-setosa
2    4.7  3.2  1.3  0.2     Iris-setosa
3    4.6  3.1  1.5  0.2     Iris-setosa
4    5.0  3.6  1.4  0.2     Iris-setosa
5    5.4  3.9  1.7  0.4     Iris-setosa
6    4.6  3.4  1.4  0.3     Iris-setosa
7    5.0  3.4  1.5  0.2     Iris-setosa
8    4.4  2.9  1.4  0.2     Iris-setosa
9    4.9  3.1  1.5  0.1     Iris-setosa
10   5.4  3.7  1.5  0.2     Iris-setosa
11   4.8  3.4  1.6  0.2     Iris-setosa
12   4.8  3.0  1.4  0.1     Iris-setosa
13   4.3  3.0  1.1  0.1     Iris-setosa
14   5.8  4.0  1.2  0.2     Iris-setosa
15   5.7  4.4  1.5  0.4     Iris-setosa
16   5.4  3.9  1.3  0.4     Iris-setosa
17   5.1  3.5  1.4  0.3     Iris-setosa
18   5.7  3.8  1.7  0.3     Iris-setosa
19   5.1  3.8  1.5  0.3     Iris-setosa
20   5.4  3.4  1.7  0.2     Iris-setosa
21   5.1  3.7  1.5  0.4     Iris-setosa
22   4.6  3.6  1.0  0.2     Iris-setosa
23   5.1  3.3  1.7  0.5     Iris-setosa
24   4.8  3.4  1.9  0.2     Iris-setosa
25   5.0  3.0  1.6  0.2     Iris-setosa
26   5.0  3.4  1.6  0.4     Iris-setosa
27   5.2  3.5  1.5  0.2     Iris-setosa
28   5.2  3.4  1.4  0.2     Iris-setosa
29   4.7  3.2  1.6  0.2     Iris-setosa
..   ...  ...  ...  ...             ...
120  6.9  3.2  5.7  2.3  Iris-virginica
121  5.6  2.8  4.9  2.0  Iris-virginica
122  7.7  2.8  6.7  2.0  Iris-virginica
123  6.3  2.7  4.9  1.8  Iris-virginica
124  6.7  3.3  5.7  2.1  Iris-virginica
125  7.2  3.2  6.0  1.8  Iris-virginica
126  6.2  2.8  4.8  1.8  Iris-virginica
127  6.1  3.0  4.9  1.8  Iris-virginica
128  6.4  2.8  5.6  2.1  Iris-virginica
129  7.2  3.0  5.8  1.6  Iris-virginica
130  7.4  2.8  6.1  1.9  Iris-virginica
131  7.9  3.8  6.4  2.0  Iris-virginica
132  6.4  2.8  5.6  2.2  Iris-virginica
133  6.3  2.8  5.1  1.5  Iris-virginica
134  6.1  2.6  5.6  1.4  Iris-virginica
135  7.7  3.0  6.1  2.3  Iris-virginica
136  6.3  3.4  5.6  2.4  Iris-virginica
137  6.4  3.1  5.5  1.8  Iris-virginica
138  6.0  3.0  4.8  1.8  Iris-virginica
139  6.9  3.1  5.4  2.1  Iris-virginica
140  6.7  3.1  5.6  2.4  Iris-virginica
141  6.9  3.1  5.1  2.3  Iris-virginica
142  5.8  2.7  5.1  1.9  Iris-virginica
143  6.8  3.2  5.9  2.3  Iris-virginica
144  6.7  3.3  5.7  2.5  Iris-virginica
145  6.7  3.0  5.2  2.3  Iris-virginica
146  6.3  2.5  5.0  1.9  Iris-virginica
147  6.5  3.0  5.2  2.0  Iris-virginica
148  6.2  3.4  5.4  2.3  Iris-virginica
149  5.9  3.0  5.1  1.8  Iris-virginica

[150 rows x 5 columns]

While we cannot do much plotting in three dimensions, it may be worth plotting two attributes at a time and exploring this dataset. 

In [2]:
import math
import collections
def dist(a, b):
    sqSum = 0
    for i in range(len(a)):
        sqSum += (a[i] - b[i]) ** 2
    return math.sqrt(sqSum)
# ------------------------------------------------ #
# We are assuming that the label is the last field #
# If not, munge the data to make it so!            #
# ------------------------------------------------ #
def kNN(k, train, given):
    distances = []
    for t in train:
        distances.append((dist(t[:-1], given), t[-1]))
    distances.sort()
    return distances[:k]

def kNN_classify(k, train, given):
    tally = collections.Counter()
    for nn in kNN(k, train, given):
        tally.update(nn[-1])
    return tally.most_common(1)[0]

**Exercise 1** :: Split the dataset into 120 + 30, for training and testing. Use K = 5 and Euclidean distance. Find the accuracy (as percentage) on the test data.

In [3]:
import pandas as pd

# Set the three iris as 0, 1, 2
# Iris Setosa -> 0
# Iris-versicolor -> 1
# Iris-virginica -> 2

def pre_process_iris(iris):
    iris = iris.strip().lower()
    if iris == 'iris-setosa':
        return 'S'
    elif iris == 'iris-versicolor':
        return 'E'
    elif iris == 'iris-virginica':
        return 'I'
    else:
        return 'X'

irisData = pd.read_csv("../../Datasets/bezdekIris.data", header=None, converters={4:pre_process_iris})
irisData.values

array([[5.1, 3.5, 1.4, 0.2, 'S'],
       [4.9, 3.0, 1.4, 0.2, 'S'],
       [4.7, 3.2, 1.3, 0.2, 'S'],
       [4.6, 3.1, 1.5, 0.2, 'S'],
       [5.0, 3.6, 1.4, 0.2, 'S'],
       [5.4, 3.9, 1.7, 0.4, 'S'],
       [4.6, 3.4, 1.4, 0.3, 'S'],
       [5.0, 3.4, 1.5, 0.2, 'S'],
       [4.4, 2.9, 1.4, 0.2, 'S'],
       [4.9, 3.1, 1.5, 0.1, 'S'],
       [5.4, 3.7, 1.5, 0.2, 'S'],
       [4.8, 3.4, 1.6, 0.2, 'S'],
       [4.8, 3.0, 1.4, 0.1, 'S'],
       [4.3, 3.0, 1.1, 0.1, 'S'],
       [5.8, 4.0, 1.2, 0.2, 'S'],
       [5.7, 4.4, 1.5, 0.4, 'S'],
       [5.4, 3.9, 1.3, 0.4, 'S'],
       [5.1, 3.5, 1.4, 0.3, 'S'],
       [5.7, 3.8, 1.7, 0.3, 'S'],
       [5.1, 3.8, 1.5, 0.3, 'S'],
       [5.4, 3.4, 1.7, 0.2, 'S'],
       [5.1, 3.7, 1.5, 0.4, 'S'],
       [4.6, 3.6, 1.0, 0.2, 'S'],
       [5.1, 3.3, 1.7, 0.5, 'S'],
       [4.8, 3.4, 1.9, 0.2, 'S'],
       [5.0, 3.0, 1.6, 0.2, 'S'],
       [5.0, 3.4, 1.6, 0.4, 'S'],
       [5.2, 3.5, 1.5, 0.2, 'S'],
       [5.2, 3.4, 1.4, 0.2, 'S'],
       [4.7, 3

In [4]:
import pandas as pd
import random

# Set the three iris as 0, 1, 2
# Iris Setosa -> 0
# Iris-versicolor -> 1
# Iris-virginica -> 2

def pre_process_iris(iris):
    iris = iris.strip().lower()
    if iris == 'iris-setosa':
        return 'S'
    elif iris == 'iris-versicolor':
        return 'E'
    elif iris == 'iris-virginica':
        return 'I'
    else:
        return 'X'

irisData = pd.read_csv("../../Datasets/bezdekIris.data", header=None, converters={4:pre_process_iris})

TRAIN_TEST_RATIO = 0.8
picker = list(range(irisData.shape[0]))
random.shuffle(picker)       ### randomly shuffle the data
trainMax = int(len(picker) * TRAIN_TEST_RATIO)
train = []
test = []
for pick in picker[:trainMax]:
    train.append(list(irisData.values[pick]))         ### select 80% of data to be used as training set
for pick in picker[trainMax:]:
    test.append(list(irisData.values[pick]))       ### select 20% of data to be used as test set

results = []
for i, t in enumerate(test):
    results.append(kNN_classify(5, train, t)[0] == test[i][-1])
print(results.count(True), "are correct")
correctPredicted = results.count(True)
totalTestData = len(test)
accuracy = (correctPredicted / totalTestData) * 100
print('The accuracy is ' + str(accuracy))

30 are correct
The accuracy is 100.0


**Exercise 2** :: Repeat the above (creating random partitions and evaluating the performance) 5 times. 

In [6]:
import pandas as pd
import random

# Set the three iris as 0, 1, 2
# Iris Setosa -> 0
# Iris-versicolor -> 1
# Iris-virginica -> 2

def pre_process_iris(iris):
    iris = iris.strip().lower()
    if iris == 'iris-setosa':
        return 'S'
    elif iris == 'iris-versicolor':
        return 'E'
    elif iris == 'iris-virginica':
        return 'I'
    else:
        return 'X'

irisData = pd.read_csv("../../Datasets/bezdekIris.data", header=None, converters={4:pre_process_iris})

j = 0
accuracyList = []
avgAccuracy = 0

for j in range(0, 5):
    TRAIN_TEST_RATIO = 0.8
    picker = list(range(irisData.shape[0]))
    random.shuffle(picker)       ### randomly shuffle the data
    trainMax = int(len(picker) * TRAIN_TEST_RATIO)
    train = []
    test = []
    for pick in picker[:trainMax]:
        train.append(list(irisData.values[pick]))         ### select 80% of data to be used as training set
    for pick in picker[trainMax:]:
        test.append(list(irisData.values[pick]))       ### select 20% of data to be used as test set

    results = []
    for i, t in enumerate(test):
        results.append(kNN_classify(5, train, t)[0] == test[i][-1])
    print(results.count(True), "are correct")
    correctPredicted = results.count(True)
    totalTestData = len(test)
    accuracy = (correctPredicted / totalTestData) * 100
    print('The accuracy is ' + str(accuracy))
    accuracyList.append(accuracy)
    avgAccuracy += accuracy
    
avgAccuracy = avgAccuracy / len(accuracyList)
print('The average accuracy is ' + str(avgAccuracy))

29 are correct
The accuracy is 96.66666666666667
30 are correct
The accuracy is 100.0
29 are correct
The accuracy is 96.66666666666667
28 are correct
The accuracy is 93.33333333333333
29 are correct
The accuracy is 96.66666666666667
The average accuracy is 96.66666666666667


**Exercise 3** :: Vary K from 3 to 11 and find the best K.

In [7]:
import pandas as pd
import random

# Set the three iris as 0, 1, 2
# Iris Setosa -> 0
# Iris-versicolor -> 1
# Iris-virginica -> 2

def pre_process_iris(iris):
    iris = iris.strip().lower()
    if iris == 'iris-setosa':
        return 'S'
    elif iris == 'iris-versicolor':
        return 'E'
    elif iris == 'iris-virginica':
        return 'I'
    else:
        return 'X'

irisData = pd.read_csv("../../Datasets/bezdekIris.data", header=None, converters={4:pre_process_iris})

j = 0
k = 3

for k in range(3, 12):
    accuracyList = []
    avgAccuracy = 0
    
    for j in range(0, 5):
        TRAIN_TEST_RATIO = 0.8
        picker = list(range(irisData.shape[0]))
        random.shuffle(picker)       ### randomly shuffle the data
        trainMax = int(len(picker) * TRAIN_TEST_RATIO)
        train = []
        test = []
        for pick in picker[:trainMax]:
            train.append(list(irisData.values[pick]))         ### select 80% of data to be used as training set
        for pick in picker[trainMax:]:
            test.append(list(irisData.values[pick]))       ### select 20% of data to be used as test set

        results = []
        for i, t in enumerate(test):
            results.append(kNN_classify(5, train, t)[0] == test[i][-1])
        print(results.count(True), "are correct")
        correctPredicted = results.count(True)
        totalTestData = len(test)
        accuracy = (correctPredicted / totalTestData) * 100
        print('The accuracy is ' + str(accuracy))
        accuracyList.append(accuracy)
        avgAccuracy += accuracy
        # Reset j
        j = 0

    avgAccuracy = avgAccuracy / len(accuracyList)
    print('The average accuracy for k = ' + str(k) + ' is ' + str(avgAccuracy))

29 are correct
The accuracy is 96.66666666666667
28 are correct
The accuracy is 93.33333333333333
29 are correct
The accuracy is 96.66666666666667
28 are correct
The accuracy is 93.33333333333333
30 are correct
The accuracy is 100.0
The average accuracy for k = 3 is 96.0
29 are correct
The accuracy is 96.66666666666667
30 are correct
The accuracy is 100.0
28 are correct
The accuracy is 93.33333333333333
29 are correct
The accuracy is 96.66666666666667
29 are correct
The accuracy is 96.66666666666667
The average accuracy for k = 4 is 96.66666666666667
29 are correct
The accuracy is 96.66666666666667
30 are correct
The accuracy is 100.0
29 are correct
The accuracy is 96.66666666666667
30 are correct
The accuracy is 100.0
30 are correct
The accuracy is 100.0
The average accuracy for k = 5 is 98.66666666666667
29 are correct
The accuracy is 96.66666666666667
30 are correct
The accuracy is 100.0
28 are correct
The accuracy is 93.33333333333333
29 are correct
The accuracy is 96.6666666666666